imap/ctl_cyrusdb.c

```diff
@@ -189,35 +191,55 @@ static int fixmbox(const mbentry_t *mbentry,
                               mbentry->name, mailbox->h.uniqueid);
         }
 
-        copy = mboxlist_entry_copy(mbentry);
+        if (!copy) mbentry = copy = mboxlist_entry_copy(mbentry);
         copy->uniqueid = xstrdup(mailbox->h.uniqueid);
         xsyslog(LOG_INFO, "mbentry had no uniqueid, setting from header",
                           "mboxname=<%s> newuniqueid=<%s>",
                           copy->name, copy->uniqueid);
 
+skip_uniqueid:
+        ;   /* hush "label at end of compound statement" warning */
+    }
+
+    if (mboxname_isusermailbox(mbentry->name, 1)) {
+        free(frock->inboxname);
+        free(frock->inboxid);
+        frock->inboxname = xstrdup(mbentry->name);
+        frock->inboxid = xstrdup(mbentry->uniqueid);
+    }
+    if (!mbentry->inboxid && frock->inboxname &&
+        mboxname_is_prefix(mbentry->name, frock->inboxname)) {
+        xsyslog(LOG_INFO, "mbentry had no inboxid, setting from",
```

Comment:
This is going to be very noisy on upgrade, and doesn't log the mbentry name.  It should probably at least do that!